heroic-dist/src/test/java/com/spotify/heroic/cluster/CoreClusterManagerIT.java

```diff
@@ -0,0 +1,216 @@
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.Test;
+
+@Slf4j
+public class CoreClusterManagerIT extends AbstractLocalClusterIT {
+    private final ControlledNodeMetadataFactory metadata1 = new ControlledNodeMetadataFactory();
+    private final ControlledNodeMetadataFactory metadata2 = new ControlledNodeMetadataFactory();
+
+    private final List<ControlledNodeMetadataFactory> metadataFactories =
+        ImmutableList.of(metadata1, metadata2);
+
+    private static final Random random = new Random();
+
+    @Override
+    protected String protocol() {
+        return "grpc";
+    }
+
+    @Override
+    protected List<NodeMetadataFactory> metadataFactories() {
+        return ImmutableList.of(metadata1, metadata2);
+    }
+
+    /**
+     * This tests the assumption that it is safe to perform refreshes, even when there are
+     * requests pending, or about to be fired.
+     *
+     * The test will continue to run, while the refresh thread is running.
+     */
+    @Test
+    public void testSafeRefresh() throws Exception {
+        final int numberOfRefreshes = 100;
+        final int requestsPerIteration = 100;
+
+        final HeroicCoreInstance a = instances.get(0);
+        final ClusterManager clusterManager = a.inject(ClusterComponent::clusterManager);
+
+        /* setup a thread that performs the given number of refresh iterations */
+        final ClusterRefreshThread t = setupRefreshThread(clusterManager, numberOfRefreshes);
+
+        final DataStreamCollector collector = new DataStreamCollector();
+
+        while (!t.shutdown.get()) {
+            final List<Callable<AsyncFuture<Void>>> operations = new ArrayList<>();
+
+            for (int i = 0; i < requestsPerIteration; i++) {
+                operations.add(() -> {
+                    final List<AsyncFuture<Void>> pongs = new ArrayList<>();
+
+                    for (final ClusterShard shard : clusterManager.useDefaultGroup()) {
+                        for (int p = 0; p < 10; p++) {
+                            final List<UUID> excludeIds = Lists.newArrayList();
+                            while (true) {
+                                // Send ping to all nodes in the shard
+                                Optional<ClusterManager.NodeResult<AsyncFuture<Void>>> ret =
+                                    clusterManager.withNodeInShardButNotWithId(shard.getShard(),
+                                        excludeIds, ClusterNode.Group::ping);
+                                if (!ret.isPresent()) {
+                                    // No more nodes available in shard
+                                    break;
```

Comment:
Now that it's in a helper I'm returning there instead.

Or did you object to the use of Optional<> as return value to indicate if there were any more nodes?